src/rgw/driver/rados/rgw_sal_rados.cc

```diff
@@ -2817,6 +2826,20 @@ int RadosMultipartUpload::complete(const DoutPrefixProvider *dpp,
 
   attrs[RGW_ATTR_ETAG] = etag_bl;
 
+  rgw_placement_rule* ru;
+  ru = &placement;
+  ret = RadosMultipartUpload::get_info(dpp, y, &ru, &attrs);
```

Comment:
i'm surprised that CompleteMultipart didn't already need access to the `multipart_upload_info`. but since this adds an extra rados read operation to CompleteMultipart that's only necessary for object lock, can we find a way to make `get_info()` conditional on something like `s->bucket->get_info().obj_lock_enabled()`?

Reply:
this appears to be overwriting the target object's `attrs` with unrelated attributes read from the multipart meta object

in testing, we see a `s3cmd get` of a multipart upload failing to return the etag header that we add above with `attrs[RGW_ATTR_ETAG] = etag_bl;`

Reply:
indeed

Reply:
i've added a commit from @mattbenjamin to address this part. i also rebased this branch to unblock testing here